### PR TITLE
vtgate: preserve IN value lists in complex aggregate projections

### PR DESCRIPTION
**Pull request by GrahamCampbell**
## Description

Planning `select count(foo) in (1, 2) from user` on a sharded keyspace panics with `hash of unhashable type sqlparser.ValTuple`. The bind variable form `select count(foo) in ::vals from user`, which is also what the normalizer rewrites literal IN lists into in production, fails with an internal error saying the rhs of an In operation should be a tuple. Both queries are valid SQL.

The root cause is in aggregate extraction. When a select expression contains an aggregate inside a larger expression, the planner extracts the aggregates and turns each remaining aggregate free subexpression into a single input column. An IN value list was captured whole this way, so offset planning replaced the entire tuple with one column reference, which can only ever carry one value. This PR teaches the extraction to treat value tuples and list bind variables as structural containers instead. Tuple members become ordinary input columns, a list bind variable stays in the expression as is, and the tuple shape survives to the expression engine, which already supports both forms.

The type lookups that panicked are also guarded. `SemTable.TypeForExpr` is a cache lookup and now reports no type for expressions that cannot be used as map keys, while `PlanningContext.TypeForExpr` still calculates the type of such expressions and only skips caching them. Planner goldens pin the literal and bind variable forms, unit tests cover both lookup contracts, and an end-to-end test compares the true, false and NULL outcomes against MySQL.

The same aggregate extraction and unguarded type lookup exist on release-23.0 and release-24.0, and both reproduce the failure: the literal form panics with `hash of unhashable type sqlparser.ValTuple` and the bind variable form fails with `rhs of an In operation should be a tuple, was *evalengine.Column`. Over the MySQL protocol the panic is a `runtime.Error` that `operators.PanicHandler` re-panics and only the per-connection handler in `go/mysql/server.go` recovers, so a client sending valid SQL has its connection closed rather than receiving an error. That makes this worth backporting: the fix is small, self-contained, and covered by planner goldens, unit tests and an end-to-end test against MySQL.

## Related Issue(s)

https://github.com/vitessio/vitess/pull/20626 depends on this change.

## Checklist

-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

## Deployment Notes

VTGate can now plan aggregate expressions compared against literal or bind variable IN lists. These queries previously failed with a planner panic or an internal error. No migrations or configuration changes.

### AI Disclosure

This PR was written primarily by Fable with review from GPT 5.6 Sol.

---

```mermaid
graph LR
    subgraph json [JSON support]
        direction TB
        PR20625["#20625 (merged)<br>mysql/json: fix MarshalTo discarding accumulated output for nested blob and bit values"]
        PR20632["#20632<br>vtgate: preserve IN value lists in complex aggregate projections"]
        PRA["#20691<br>evalengine: disable the static IN hash table for JSON operands"]
        PR20682["#20682 (draft)<br>evalengine: support constant-folded JSON values as literals"]
        PR20683["#20683 (draft)<br>evalengine, sqlparser: MySQL comparison domains; nested BETWEEN parentheses"]
        PR20626["#20626 (draft)<br>vtgate: support cross-shard JSON_ARRAYAGG and JSON_OBJECTAGG"]
        PR20625 --> PR20682
        PRA --> PR20682
        PR20682 --> PR20683
        PR20632 --> PR20626
        PR20683 --> PR20626
    end

    subgraph union [Union routing]
        direction TB
        PR20628["#20628<br>vtgate: track per-source copies of pushed join predicates so merges skip them"]
        PR20629["#20629 (merged)<br>vtgate: fix None routing handling when merging unions"]
        PR20630["#20630 (draft)<br>vtgate: merge unions on join-predicate-free routings when all sources agree"]
        PR20755["#20755<br>vtgate: fix union merging through reference-table alternates"]
        PR20756["#20756 (draft)<br>vtgate: merge empty reference branches through rewritten copies"]
        PR20628 --> PR20630
        PR20629 --> PR20630
        PR20630 --> PR20756
        PR20755 --> PR20756
    end

    subgraph cte [Recursive CTEs]
        direction TB
        PR20759["#20759<br>vtgate: expand qualified stars without JOIN USING coalescing"]
        PR20633["#20633 (draft)<br>vtgate: validate the declared column list length of CTEs and derived tables"]
        PR20631["#20631 (draft)<br>vtgate: bind recursive CTE column filters as arguments in unmerged term queries"]
        PR20759 --> PR20633
        PR20633 --> PR20631
    end

    json ~~~ union ~~~ cte
```

### CODE REVIEW

File: go/vt/vtgate/planbuilder/operators/queryprojection.go [hunk 1]
```diff
@@ -481,6 +481,14 @@ func (qp *QueryProjection) extractAggr(
 			makeComplex()
 			return true
 		}
+		switch ex.(type) {
+		case sqlparser.ValTuple, sqlparser.ListArg:
+			// Tuples are structural containers, not values a single input
+			// column could carry: an IN value list must stay a tuple all the
+			// way to the evalengine. Descend so the members become ordinary
+			// inputs; a list bind variable stays in the expression as is.
```
[the remark this answers is not in the document]
Comment (GrahamCampbell) [reply go/vt/vtgate/planbuilder/operators/queryprojection.go]: Trimmed these in the restack: the tuple comment is now a single sentence and both ValidAsMapKey guards are one-liners matching the existing wording elsewhere in semantic_table.go — kept as short notes rather than deleted outright, since the guard's reason isn't obvious from the call site.

**Review by migue — COMMENTED** [review]
Reviewed this one by checking out the branch and running it rather than just reading the diff. The fix looks right to me, and it is consistent with how this code already behaves.

What I verified:

- The panic reproduces on main (`fd0e84b28e`) for the literal form, as does `rhs of an In operation should be a tuple, was *evalengine.Column` for `::vals`. Both plan correctly on this branch, matching the goldens exactly.
- `planbuilder`, `semantics` and `evalengine` suites pass 
- Twelve neighbouring cases plan sensibly here and panic or error on main: `NOT IN`, a tuple containing a column, a tuple containing another aggregate, nested row constructors `(count(foo), 1) in ((1, 2), (3, 4))`, `CASE WHEN`, `GROUP BY`, `HAVING`, `DISTINCT`, `ORDER BY`, and `IN ::vals` with `GROUP BY`.
- The projection expression evaluates correctly at runtime on the compiled evalengine path, three-valued logic included.

One thing worth raising: this is a bit worse than a planning error. `operators.PanicHandler` re-panics on `runtime.Error`, and `hash of unhashable type` is one. Over gRPC `VTGate.HandlePanic` catches it, but over the MySQL protocol it unwinds to `go/mysql/server.go:394`, which logs and closes the connection, so today a client sending valid SQL gets a dropped connection rather than an error. That reads like a backport argument to me?

File: go/vt/vtgate/semantics/semantic_table.go [hunk 2]
```diff
@@ -673,6 +673,10 @@ func (st *SemTable) AddExprs(tbl *sqlparser.AliasedTableExpr, cols *sqlparser.Se
 // TypeForExpr returns the type of expressions in the query
 // Note that PlanningContext has the same method, and you should use that if you have a PlanningContext
 func (st *SemTable) TypeForExpr(e sqlparser.Expr) (evalengine.Type, bool) {
+	if !ValidAsMapKey(e) {
```
Comment (migue) [inline go/vt/vtgate/semantics/semantic_table.go:676]: `ValidAsMapKey` is `reflect.TypeOf(s).Comparable()`, and `reflect.TypeOf` on a nil interface returns `nil`, so this guard panics on a nil `Expr` where the old code returned cleanly?

```go
var e sqlparser.Expr
st.TypeForExpr(e)
// main: typ=Unknown, found=false
// here: panic: runtime error: invalid memory address or nil pointer dereference
```

(A nil interface is itself a valid map key, so `st.ExprTypes[nil]` was always fine.)

I couldn't find a live caller (the nil sources I traced, `Aggr.GetTypeCollation` and `NeedsWeightString`, both call `aggr.Func.GetArg()`, and `*CountStar.GetArg()` returns nil, are gated behind `NeedsComparableValues()`, which excludes `CountStar`). So this is latent rather than a regression anyone hits today. But the guard exists precisely to harden this lookup, and `TypeForExpr` is called from everywhere:

```go
if e == nil || !ValidAsMapKey(e) {
```

Or push it into `ValidAsMapKey` itself, since the other five call sites share the hazard.

File: go/vt/vtgate/planbuilder/operators/queryprojection.go [hunk 3]
```diff
@@ -481,6 +481,13 @@ func (qp *QueryProjection) extractAggr(
 			makeComplex()
 			return true
 		}
+		switch ex.(type) {
+		case sqlparser.ValTuple, sqlparser.ListArg:
```
Comment (migue) [inline go/vt/vtgate/planbuilder/operators/queryprojection.go:485]: Small thing on the comment: it reads as one rule, but there are two mechanisms here and only the first matches the wording.

`ValTuple` really is a container, `return true` descends, so each member becomes its own input column and the tuple shape survives. `ListArg` is a leaf (`type ListArg string`), so `return true` has no children to visit: it means "fall through without capturing this", and the bind variable is left for the evalengine to resolve at runtime.

Read cold, the `ListArg` half looks like an accidental no-op, which is the kind of thing a later refactor collapses into `return false`. Maybe:

```go
// A tuple is a structural container, not a value a single input column
// could carry: descend, so each member becomes its own column and the
// tuple shape survives to the evalengine. A list bind variable is a leaf
// that must simply be left in place — the evalengine resolves it at runtime.
```

File: go/vt/vtgate/planbuilder/testdata/aggr_cases.json [hunk 4]
```diff
@@ -7408,5 +7408,79 @@
         "user.user"
       ]
     }
+  },
+  {
+    "comment": "vtgate-evaluated aggregate compared against a literal IN value list keeps the tuple structure",
```
Comment (migue) [inline go/vt/vtgate/planbuilder/testdata/aggr_cases.json:7413]: These goldens pin that a plan is *produced*, but nothing here pins that it *computes the right answer*, which is the actual claim in the description ("the tuple shape survives to the expression engine, which already supports both forms").

I checked it by hand against the post-offset-planning shape, and it is correct on the compiled path, NULL semantics included:

| row | result |
| --- | --- |
| `2 IN (1, 2)` | `1` |
| `5 IN (1, 2)` | `0` |
| `5 IN (1, NULL)` | `NULL` |

plus the `ListArg` form with a `TUPLE` bind variable. Would you consider one `mcmp.Exec` case in `go/test/endtoend/vtgate/queries/aggregation/aggregation_test.go`? It would compare against real MySQL and cover the NULL case, which is the one most likely to drift.

**Review by migue — APPROVED** [review]
Thanks a lot for taking a look at my comments! I am approving it but it would be worth that someone with deeper knowledge in the area take a look at this

**Review by maxenglander — APPROVED** [review]
lgtm, thanks @GrahamCampbell. please address all the open threads from @migue 🙏

**Comment by maxenglander** [comment]
@GrahamCampbell i added backport labels, but i'm not sure if this bug actually impacts v23 and v24. would you mind validating that? if so, you can update the PR description to mention why we are backporting and tick off the two backport-related checkboxes.

File: go/vt/vtgate/semantics/semantic_table.go (hunk 2 above)
Comment (GrahamCampbell) [reply go/vt/vtgate/semantics/semantic_table.go:676]: Right, `reflect.TypeOf(nil)` is a nil `reflect.Type` and calling `Comparable()` on it panics, where the old map lookup happily used the nil key. I took the second option: `ValidAsMapKey` now returns true for a nil node (a nil interface is a valid map key), so all six call sites are covered rather than just this one, and `TestTypeForExprUnhashable` also asserts that `TypeForExpr(nil)` reports no type. Agreed it is latent today; the `CountStar.GetArg()` sources you traced are gated the way you describe.

File: go/vt/vtgate/planbuilder/operators/queryprojection.go (hunk 3 above)
Comment (GrahamCampbell) [reply go/vt/vtgate/planbuilder/operators/queryprojection.go:485]: Good point, the old wording only described the tuple half. Reworded to cover both mechanisms: the tuple is descended so each member becomes its own input column and the tuple shape survives to the evalengine, and the list bind variable is a leaf that is deliberately left in place for the evalengine to resolve at runtime, so the `return true` on `ListArg` no longer reads like an accidental no-op.

File: go/vt/vtgate/planbuilder/testdata/aggr_cases.json (hunk 4 above)
Comment (GrahamCampbell) [reply go/vt/vtgate/planbuilder/testdata/aggr_cases.json:7413]: Added `TestAggregateInValueList` in `go/test/endtoend/vtgate/queries/aggregation/aggregation_test.go`. It seeds `aggr_test` with four rows (one NULL `val2`) and compares vtgate with MySQL for `count(*) in (3, 4)` (1), `count(*) in (1, 2)` (0), `count(val2) in (3, null)` (1), `count(val2) in (1, null)` (NULL) and `count(*) not in (1, null)` (NULL), so the three-valued cases are pinned against real MySQL. Through vtgate the literal lists are normalized into a list bind variable, so the bind-variable path is what actually executes there; the literal form stays covered by the planner goldens.

**Comment by GrahamCampbell** [comment]
Validated: both release-23.0 (`e98e0cda8c`) and release-24.0 (`8915fde898`) carry the same aggregate extraction and unguarded type lookup, and both reproduce through the planner: `select count(foo) in (1, 2) from user` panics with `hash of unhashable type sqlparser.ValTuple` and `select count(foo) in ::vals from user` fails with `rhs of an In operation should be a tuple, was *evalengine.Column`. Over the MySQL protocol the panic is only recovered by the per-connection handler in `go/mysql/server.go`, which logs it and closes the connection. I have added the justification to the description and ticked the two backport boxes, and the latest push addresses all three of @migue's threads.

**Review by migue — APPROVED** [review]
